cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/lib/table/ObjectMappedTableDataset.java

```diff
@@ -0,0 +1,216 @@
+/*
+ * Copyright © 2015 Cask Data, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package co.cask.cdap.data2.dataset2.lib.table;
+
+import co.cask.cdap.api.annotation.Beta;
+import co.cask.cdap.api.common.Bytes;
+import co.cask.cdap.api.data.batch.RecordScanner;
+import co.cask.cdap.api.data.batch.Split;
+import co.cask.cdap.api.data.batch.SplitReader;
+import co.cask.cdap.api.data.format.StructuredRecord;
+import co.cask.cdap.api.data.schema.Schema;
+import co.cask.cdap.api.dataset.DataSetException;
+import co.cask.cdap.api.dataset.lib.AbstractCloseableIterator;
+import co.cask.cdap.api.dataset.lib.AbstractDataset;
+import co.cask.cdap.api.dataset.lib.CloseableIterator;
+import co.cask.cdap.api.dataset.lib.KeyValue;
+import co.cask.cdap.api.dataset.lib.ObjectMappedTable;
+import co.cask.cdap.api.dataset.table.Put;
+import co.cask.cdap.api.dataset.table.Row;
+import co.cask.cdap.api.dataset.table.Scanner;
+import co.cask.cdap.api.dataset.table.Table;
+import co.cask.cdap.internal.io.ReflectionPutWriter;
+import co.cask.cdap.internal.io.ReflectionRowReader;
+import co.cask.cdap.internal.io.TypeRepresentation;
+import com.google.common.base.Preconditions;
+import com.google.common.reflect.TypeToken;
+
+import java.io.IOException;
+import java.lang.reflect.Type;
+import java.util.List;
+import javax.annotation.Nullable;
+
+/**
+ * Default implementation for {@link ObjectMappedTable}.
+ *
+ * @param <T> the type of objects in the table
+ */
+@Beta
+public class ObjectMappedTableDataset<T> extends AbstractDataset implements ObjectMappedTable<T> {
+
+  private final Table table;
+  private final Schema schema;
+
+  private final ReflectionPutWriter<T> putWriter;
+  private final ReflectionRowReader<T> rowReader;
+
+  @SuppressWarnings("unchecked")
+  public ObjectMappedTableDataset(String name, Table table, TypeRepresentation typeRep,
```

Comment:
ok actually, because of the classloading issue described in the previous comment, since we have the schema in the definition it is simpler to just pass it in instead of generating it from the type each time a table is constructed